brainframe_qt/util/oauth/cognito.py

```diff
@@ -0,0 +1,207 @@
+import base64
+import hashlib
+import json
+import secrets
+import string
+import uuid
+from json import JSONDecodeError
+from typing import Tuple
+
+from PyQt5.QtCore import QObject, pyqtSignal, QUrl, QUrlQuery
+from PyQt5.QtNetwork import QNetworkRequest, QNetworkAccessManager, QNetworkReply
+
+from brainframe.api import bf_codecs
+
+from brainframe_qt import constants
+from brainframe_qt.util.oauth.base import AuthTokenResponse
+from brainframe_qt.util.oauth.reply_handler import PKCEReplyHandler
+
+
+class CognitoOAuth(QObject):
+    """Performs OAuth2 Authorization Code Flow with PKCE with Aotu's Cognito instance"""
+
+    ready_to_authenticate = pyqtSignal(QUrl)
+    """Emitted when the Authentication URL is ready to be opened in a browser window
+    on the client computer for authentication"""
+
+    authentication_successful = pyqtSignal(bf_codecs.CloudTokens)
+    """Emitted when Authentication is successful and both Access and Refresh Tokens have
+    been acquired"""
+
+    authentication_error = pyqtSignal(Exception)
+    """Emitted when there was an Error during the Authentication procedure"""
+
+    class InvalidStateError(Exception):
+        """State returned to redirect_uri did not match state sent to auth server"""
+
+    class InvalidTokenResponse(Exception):
+        """Reply to Access and Refresh Tokens request was invalid"""
+
+    _SCOPES = ["email", "aws.cognito.signin.user.admin", "profile", "openid",
+               *constants.oauth.OAUTH_SCOPES]
```

Comment:
uhh great question

Reply:
I think it was because the constants in `constants.oauth` are specific to _Aotu_'s implementation of OAuth and not generic ones that would be used by any OAuth (or even Cognito) service